source/yats/include/yats/pipeline.h

```diff
@@ -40,47 +44,222 @@ class pipeline
+        file << '\t' << "node [shape = record];" << std::endl;
+
+        auto helpers = get_helpers();
+        auto output_owners = get_output_owners(helpers);
+
+        // Creates a vertex for each task including the task's inputs and outputs.
+        // The tasks are named n0, n1, ... according to their helper index. They are labeled with their class name.
+        // Key for input values are i1, i2, .... Keys for out values are o1, o2, ... .
+        // NODE_NAME [label = "NODE_NAME|{{<KEY1>INPUT1|<KEY2>INPUT2...}|{<KEY3>OUTPUT1|<KEY4>OUTPUT2...}}"];
+        for (size_t i = 0; i < helpers.size(); ++i)
+        {
+            file << '\t' << "n" << i << "[label = \"" << helpers[i]->get_task_name() << "|{";
+            file << '{' << inputs_to_string(*helpers[i]) << "}|";
+            file << '{' << outputs_to_string(*helpers[i]) << '}';
+            file << "}\"]" << std::endl;
+        }
+
+        file << std::endl;
+
+        // We need to keep track of outputs which are not connected.
+        std::set<const abstract_output_connector*> unused_outputs;
         for (size_t i = 0; i < m_tasks.size(); ++i)
         {
             auto outputs = helpers[i]->outputs();
-            for (auto output : outputs)
+            for (const auto output : outputs)
             {
-                output_owner.emplace(output.first, i);
+                unused_outputs.insert(output.first);
             }
         }
 
+        // Counter to name either inputs not fed by an output or outputs not connected by an
+        // input.
+        auto id_counter = 0;
+
         for (size_t i = 0; i < helpers.size(); ++i)
         {
             auto inputs = helpers[i]->inputs();
-            for (auto input : inputs)
+            for (const auto input : inputs)
             {
                 auto source_location = input.first->output();
-                auto source_task_id = output_owner.at(source_location);
 
-                helpers[source_task_id]->bind(source_location, helpers[i]->target(input.first));
-                helpers[source_task_id]->add_following(i);
+                // Input is not connected to an output
+                if (source_location == nullptr)
+                {
+                    file << '\t' << "node [shape = point]; ";
+                    file << 'u' << id_counter << ';' << std::endl;
+                    file << '\t' << 'u' << id_counter << "->" << 'n' << i << ':' << "<i" << input.second << '>' << std::endl;
+                    ++id_counter;
+                } // normal case. Adds edge from the output connected to the current input.
+                else
+                {
+                    const auto source_task_id = output_owners.at(source_location);
+                    file << '\t' << 'n' << source_task_id << ':' << "<o" << helpers[source_task_id]->get_output_index(source_location) << "> -> " << 'n' << i << ':' << "<i" << input.second << '>' << std::endl;
+                    unused_outputs.erase(source_location);
```

Comment:
instead of adding every output to this set and removing most of them here again it would be better to just insert them in the other if branch.

Reply:
Yes, but then we would have to calculate the difference between all outputs and the used outputs.